components/style/values/specified/mod.rs

```diff
@@ -655,21 +654,15 @@ impl Shadow {
             if !lengths_parsed {
                 if let Ok(value) = input.try(|i| Length::parse(context, i)) {
                     lengths[0] = value;
-                    let mut length_parsed_count = 1;
-                    while length_parsed_count < length_count {
-                        if let Ok(value) = input.try(|i| Length::parse(context, i)) {
-                            lengths[length_parsed_count] = value
-                        } else {
-                            break
+                    lengths[1] = try!(Length::parse(context, input));
+                    if let Ok(value) = input.try(|i| Length::parse_non_negative(i)) {
+                        lengths[2] = value;
+                        if !disable_spread_and_inset {
```

Comment:
You can add something like
```rust
debug_assert!(!disable_spread_and_inset || lengths[3] == Length::zero());
```
before returning from the function.

This assertion would run in debug build, so that if this invariant gets broken, we can know it easier without diagnosing too much.